File: go/1219.go
func dfs(x, y int, grid [][]int) int {
    n := len(grid)
    m := len(grid[0])
    maxx := 0
    dir := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
    tmp := grid[x][y]
    grid[x][y] = 0
    for _, d := range dir {
        nx := x + d[0]
        ny := y + d[1]
        if nx < 0 || ny < 0 || nx >= n || ny >= m || grid[nx][ny] == 0 {
            continue
        }
        res := dfs(nx, ny, grid)
        if res > maxx {
            maxx = res
        }
    }
    grid[x][y] = tmp
    return maxx + tmp
}

func getMaximumGold(grid [][]int) int {
    n := len(grid)
    m := len(grid[0])
    ans := 0
    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            if grid[i][j] != 0 {
                res := dfs(i, j, grid)
                if res > ans {
                    ans = res
                }
            }
        }
    }
    return ans
}

File: rust/1219.rs
impl Solution {
    pub fn get_maximum_gold(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let m = grid[0].len();

        fn dfs(x: usize, y: usize, grid: &mut Vec<Vec<i32>>) -> i32 {
            let dir = vec![(-1, 0), (1, 0), (0, -1), (0, 1)];
            let mut maxx = 0;
            let n = grid.len();
            let m = grid[0].len();

            let tmp = grid[x][y];
            grid[x][y] = 0;
            for d in dir {
                let nx = (x as i32 + d.0) as usize;
                let ny = (y as i32 + d.1) as usize;
                if nx >= n || ny >= m || grid[nx][ny] == 0 {
                    continue;
                }
                maxx = std::cmp::max(maxx, dfs(nx, ny, grid));
            }
            grid[x][y] = tmp;
            return maxx + grid[x][y];
        }

        let mut ans = 0;
        let mut grid = grid;
        for i in 0..n {
            for j in 0..m {
                if grid[i][j] != 0 {
                    ans = std::cmp::max(ans, dfs(i, j, &mut grid));
                }
            }
        }
        ans
    }
}
